plan/aggregation_pruning.go

```
@@ -81,22 +71,40 @@ func (ap *aggPruner) eliminateAggregation(p LogicalPlan) (LogicalPlan, error) {
 	return retPlan, nil
 }
 
+func convertAggToProj(agg *Aggregation, ctx context.Context, allocator *idAllocator) (*Projection, error) {
+	proj := &Projection{
+		Exprs:           make([]expression.Expression, 0, len(agg.AggFuncs)),
+		baseLogicalPlan: newBaseLogicalPlan(Proj, allocator),
+	}
+	proj.self = proj
+	proj.initIDAndContext(ctx)
+	for _, fun := range agg.AggFuncs {
+		expr, err := rewriteExpr(fun.GetArgs(), fun.GetName(), ctx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		proj.Exprs = append(proj.Exprs, expr)
+	}
+	proj.SetSchema(agg.schema.Clone())
+	return proj, nil
+}
+
 // rewriteExpr will rewrite the aggregate function to expression doesn't contain aggregate function.
-func (ap *aggPruner) rewriteExpr(exprs []expression.Expression, funcName string) (newExpr expression.Expression, err error) {
+func rewriteExpr(exprs []expression.Expression, funcName string, ctx context.Context) (newExpr expression.Expression, err error) {
 	switch funcName {
 	case ast.AggFuncCount:
 		// If is count(expr), we will change it to if(isnull(expr), 0, 1).
 		// If is count(distinct x, y, z) we will change it to if(isnull(x) or isnull(y) or isnull(z), 0, 1).
 		isNullExprs := make([]expression.Expression, 0, len(exprs))
 		for _, expr := range exprs {
-			isNullExpr, err := expression.NewFunction(ap.ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), expr.Clone())
+			isNullExpr, err := expression.NewFunction(ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), expr.Clone())
```

We can ignore this err, because it never be not nil.